File: ZWWebViewHandler/ZWEventHandler/ZWEventHandler.m
//
//  ZWEventHandler.m
//  ZWWebViewHandler
//
//  Created by zzw on 2018/7/17.
//  Copyright © 2018年 zzw. All rights reserved.
//

#import "ZWEventHandler.h"
#if DEBUG
#define ZWLog(FORMAT, ...) fprintf(stderr,"\nfunction:%s line:%d content:%s\n", __FUNCTION__, __LINE__, [[NSString stringWithFormat:FORMAT, ##__VA_ARGS__] UTF8String]);
#else
#define ZWLog(FORMAT, ...) nil
#endif
 NSString * const EventHandler = @"ZWEventHandler";

@interface ZWEventHandlerEmptyObject :NSObject

@end

@implementation ZWEventHandlerEmptyObject

@end

@implementation ZWEventHandler

+ (instancetype)instance{

    ZWEventHandler * handler  = [[self alloc] init];
    handler.handlerJS = [handler getJsString];
    return handler;

}
- (NSString *)getJsString{

    NSString *path =[[NSBundle bundleForClass:[self class]] pathForResource:@"ZWEventHandler" ofType:@"js"];
    NSString *handlerJS = [NSString stringWithContentsOfFile:path encoding:kCFStringEncodingUTF8 error:nil];
    handlerJS = [handlerJS stringByReplacingOccurrencesOfString:@"\n" withString:@""];
    return handlerJS;
}

#pragma mark - WKScriptMessageHandler
- (void)userContentController:(WKUserContentController *)userContentController
      didReceiveScriptMessage:(WKScriptMessage *)message {
    // NSLog(@"message :%@",message.body);
#pragma clang diagnostic push
#pragma clang diagnostic ignored"-Wincompatible-pointer-types-discards-qualifiers"
    if ([message.name isEqualToString:EventHandler]) {
#pragma clang diagnostic pop
        NSString *methodName = message.body[@"methodName"];
        NSDictionary *params = message.body[@"params"];

        NSString *type = message.body[@"type"];
        if ([type isEqualToString:@"CallBackFunction"]) {
            NSString *successCallBackID = message.body[@"successCallBackID"];
            NSString *failureCallBackID = message.body[@"failureCallBackID"];
            __weak typeof(self) weakSelf = self;
            [self newInteractWitMethodName:methodName params:params success:^(id response) {
                [weakSelf _zwCallJSCallBackWithCallBackName:successCallBackID response:response];
            } failure:^(id response) {
                [weakSelf _zwCallJSCallBackWithCallBackName:failureCallBackID response:response];
            }];
        }else{

            NSString *callBackName = message.body[@"callBackID"];
            if (callBackName) {
                __weak typeof(self) weakSelf = self;
                [self interactWitMethodName:methodName params:params :^(id response) {

                    [weakSelf _zwCallJSCallBackWithCallBackName:callBackName response:response];
                }];
            }else{
                [self interactWitMethodName:methodName params:params :nil];
            }
        }

    }


}


- (void)interactWitMethodName:(NSString *)methodName params:(NSDictionary *)params :(void(^)(id response))callBack{

    if (params) {
        methodName = [NSString stringWithFormat:@"%@:",methodName];
        if (callBack) {
            methodName = [NSString stringWithFormat:@"%@:",methodName];
            SEL selector =NSSelectorFromString(methodName);
            NSArray *paramArray =@[params,callBack];
            if ([self.target respondsToSelector:selector]) {
                [self _zwPerformSelector:selector withObjects:paramArray];
            }
        }else{
            SEL selector =NSSelectorFromString(methodName);
            NSArray *paramArray =@[params];
            if ([self.target respondsToSelector:selector]) {
                NSLog(@"是否调用");
                [self _zwPerformSelector:selector withObjects:paramArray];

            }
        }
    }else{

        if (callBack) {
            methodName = [NSString stringWithFormat:@"%@:",methodName];
            SEL selector =NSSelectorFromString(methodName);
            NSArray *paramArray =@[callBack];
            if ([self.target respondsToSelector:selector]) {
                [self _zwPerformSelector:selector withObjects:paramArray];
            }
        }else{
            SEL selector =NSSelectorFromString(methodName);

            if ([self.target respondsToSelector:selector]) {

                [self _zwPerformSelector:selector withObjects:nil];

            }
        }
    }
}

- (void)newInteractWitMethodName:(NSString *)methodName params:(NSDictionary *)params success:(void(^)(id response))successCallBack failure:(void(^)(id response))failureCallBack{
    if (params) {
        methodName = [NSString stringWithFormat:@"%@:",methodName];

        methodName = [NSString stringWithFormat:@"%@::",methodName];
        SEL selector =NSSelectorFromString(methodName);
        id successBlock=nil;
        if (successCallBack) {
            successBlock = successCallBack;
        }else{
            successBlock = [ZWEventHandlerEmptyObject class];
        }

        id failureBlock=nil;
        if (failureCallBack) {
            failureBlock = failureCallBack;
        }else{
            failureBlock = [ZWEventHandlerEmptyObject class];
        }
        NSArray *paramArray =@[params,successBlock,failureBlock];
        if ([self.target respondsToSelector:selector]) {
            [self _zwPerformSelector:selector withObjects:paramArray];
        }


    }
}

- (id)_zwPerformSelector:(SEL)aSelector withObjects:(NSArray *)objects {

    if (self.target) {

        NSMethodSignature *signature = [self.target methodSignatureForSelector:aSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setTarget:self.target];
        [invocation setSelector:aSelector];

        NSUInteger i = 1;

        for (id object in objects) {
            id tempObject = object;
            if (![tempObject isKindOfClass:[NSObject class]]) {
                if ([tempObject isSubclassOfClass:[ZWEventHandlerEmptyObject class]]) {
                    tempObject = nil;
                }
            }
            [invocation setArgument:&tempObject atIndex:++i];
        }
        [invocation invoke];

        if ([signature methodReturnLength]) {
            id data;
            [invocation getReturnValue:&data];
            return data;
        }
    }

    return nil;
}

- (void)_zwCallJSCallBackWithCallBackName:(NSString *)callBackName response:(id)response{
    __weak  WKWebView *weakWebView = _webView;
    NSString *js = [NSString stringWithFormat:@"ZWEventHandler.callBack('%@','%@');",callBackName,response];
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakWebView evaluateJavaScript:js completionHandler:^(id _Nullable data, NSError * _Nullable error) {
            ZWLog(@"ZWEventHandler.callBack: %@\n response: %@",callBackName,response);
        }];
    });
}


@end

File: ZWWebViewHandler/ZWEventHandler/Resources/ZWEventHandler.js
var ZWEventHandler ={
    
callNativeFunction:function(nativeMethodName,params,callBackID,callBack){
    var message;
    
    if(!callBack){
        
        message = {'methodName':nativeMethodName,'params':params};
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
        
    }else{
        message = {'methodName':nativeMethodName,'params':params,'callBackID':callBackID};
        if(!ZWBridgeEvent._listeners[callBackID]){
        ZWBridgeEvent.addEvent(callBackID, function(data){
                       
                       callBack(data);
                       
                       });
        }
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
    }
    
    
},
newCallNativeFunction:function(nativeMethodName,params,callBackID,successCallBack,failureCallBack){
    var message;
    if(successCallBack && failureCallBack){
        var successCallBackID = callBackID;
        successCallBackID +='successCallBack';
        
        var failureCallBackID = callBackID;
        failureCallBackID +='failureCallBack';
        
        message = {'type':'CallBackFunction','methodName':nativeMethodName,'params':params,'successCallBackID':successCallBackID,'failureCallBackID':failureCallBackID};
        if(!ZWBridgeEvent._listeners[successCallBackID]){
            ZWBridgeEvent.addEvent(successCallBackID, function(data){
                           
                           successCallBack(data);
                           
                           });
        }
        if(!ZWBridgeEvent._listeners[failureCallBackID]){
            ZWBridgeEvent.addEvent(failureCallBackID, function(data){
                           
                           failureCallBack(data);
                           
                           });
        }
        
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
        
    }else if(successCallBack && !failureCallBack){
        var successCallBackID = callBackID;
        successCallBackID +='successCallBack';
        message = {'type':'CallBackFunction','methodName':nativeMethodName,'params':params,'successCallBackID':successCallBackID};
        if(!ZWBridgeEvent._listeners[successCallBackID]){
            ZWBridgeEvent.addEvent(successCallBackID, function(data){
                           
                           successCallBack(data);
                           
                           });
        }
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
    }else if(failureCallBack && !successCallBack){
        var failureCallBackID = callBackID;
        failureCallBackID +='failureCallBack';
        message = {'type':'CallBackFunction','methodName':nativeMethodName,'params':params,'failureCallBackID':failureCallBackID};
        if(!ZWBridgeEvent._listeners[failureCallBackID]){
            ZWBridgeEvent.addEvent(failureCallBackID, function(data){
                           
                           failureCallBack(data);
                           
                           });
        }
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
    }
    else{
        message = {'type':'CallBackFunction','methodName':nativeMethodName,'params':params};
        window.webkit.messageHandlers.ZWEventHandler.postMessage(message);
    }
},
    
callBack:function(callBackID,data){
    
    ZWBridgeEvent.fireEvent(callBackID,data);
    
},
    
removeAllCallBacks:function(data){
    ZWBridgeEvent._listeners ={};
}
    
};



var ZWBridgeEvent = {
    
_listeners: {},
    
    
addEvent: function(type, fn) {
    if (typeof this._listeners[type] === "undefined") {
        this._listeners[type] = [];
    }
    if (typeof fn === "function") {
        this._listeners[type].push(fn);
    }
    
    return this;
},
    
    
fireEvent: function(type,param) {
    var arrayEvent = this._listeners[type];
    if (arrayEvent instanceof Array) {
        for (var i=0, length=arrayEvent.length; i<length; i+=1) {
            if (typeof arrayEvent[i] === "function") {
                arrayEvent[i](param);
            }
        }
    }
    
    return this;
},
    
removeEvent: function(type, fn) {
    var arrayEvent = this._listeners[type];
    if (typeof type === "string" && arrayEvent instanceof Array) {
        if (typeof fn === "function") {
            for (var i=0, length=arrayEvent.length; i<length; i+=1){
                if (arrayEvent[i] === fn){
                    this._listeners[type].splice(i, 1);
                    break;
                }
            }
        } else {
            delete this._listeners[type];
        }
    }
    
    return this;
}
};


